modules/openthread/platform/hdlc_interface.cpp

```diff
@@ -0,0 +1,404 @@
+	, mReceiveFrameContext(nullptr)
+	, mHdlcSpinelDecoder()
+	, mIsInitialized(false)
+	, mSavedFrame(nullptr)
+	, mSavedFrameLen(0)
+	, mIsSpinelBufferFull(false)
+	, mRadioUrl(aRadioUrl)
+{
+	mHdlcRxCallbackField = HdlcRxCallback;
+
+	radio_dev = DEVICE_DT_GET(DT_CHOSEN(zephyr_hdlc_rcp_if));
+	__ASSERT_NO_MSG(device_is_ready(radio_dev));
+
+	hdlc_api = (struct hdlc_api *)radio_dev->api;
+	if (!hdlc_api)
+	{
+		otLogDebgPlat("Radio device initialization failed");
+		assert(0);
+	}
+}
+
+HdlcInterface::~HdlcInterface(void)
+{
+}
+
+otError HdlcInterface::Init(ReceiveFrameCallback aCallback, void *aCallbackContext, RxFrameBuffer &aFrameBuffer)
+{
+	int status;
+	otError error = OT_ERROR_NONE;
+
+	if (!mIsInitialized)
+	{
+		/* Event initialization */
+		k_event_init(&s_spinel_hdlc_event);
+
+		/* Read/Write semaphores initialization */
+		k_mutex_init(&s_spinel_hdlc_wr_lock);
+		k_mutex_init(&s_spinel_hdlc_rd_lock);
+
+		if (error == OT_ERROR_NONE)
+		{
+			/* Message queue initialization */
+			k_msgq_init(&spinel_hdlc_msgq, &spinel_hdlc_msgq_buffer, 1, 1);
+
+			mHdlcSpinelDecoder.Init(mRxSpinelFrameBuffer, HandleHdlcFrame, this);
+			mReceiveFrameCallback = aCallback;
+			mReceiveFrameContext = aCallbackContext;
+			mReceiveFrameBuffer = &aFrameBuffer;
+
+			/* Initialize the HDLC interface */
+			status = hdlc_api->register_rx_cb(mHdlcRxCallbackField, this);
+			if (status == 0)
+			{
+				mIsInitialized = true;
+			}
+			else
+			{
+				otLogDebgPlat("Failed to initialize HDLC interface %d", status);
+				error = OT_ERROR_FAILED;
+			}
+		}
+	}
+
+	return error;
+}
+
+void HdlcInterface::Deinit(void)
+{
+	int status;
+
+	status = hdlc_api->deinit();
+	if (status == 0)
+	{
+		mIsInitialized = false;
+	}
+	else
+	{
+		otLogDebgPlat("Failed to terminate HDLC interface %d", status);
+	}
+
+	mReceiveFrameCallback = nullptr;
+	mReceiveFrameContext  = nullptr;
+	mReceiveFrameBuffer   = nullptr;
+}
+
+void HdlcInterface::Process(const void *aInstance)
+{
+	OT_UNUSED_VARIABLE(aInstance);
+	TryReadAndDecode(false);
+}
+
+otError HdlcInterface::SendFrame(const uint8_t *aFrame, uint16_t aLength)
+{
+	otError error = OT_ERROR_NONE;
+
+	/* Protect concurrent Send operation to avoid any buffer corruption */
+	if (k_mutex_lock(&s_spinel_hdlc_wr_lock, K_FOREVER) != 0)
+	{
+		error = OT_ERROR_FAILED;
+		goto exit;
+	}
+
+	assert(mEncoderBuffer.IsEmpty());
+
+	SuccessOrExit(error = mHdlcEncoder.BeginFrame());
+	SuccessOrExit(error = mHdlcEncoder.Encode(aFrame, aLength));
+	SuccessOrExit(error = mHdlcEncoder.EndFrame());
+	otLogDebgPlat("frame len to send = %d/%d", mEncoderBuffer.GetLength(), aLength);
+	SuccessOrExit(error = Write(mEncoderBuffer.GetFrame(), mEncoderBuffer.GetLength()));
+
+exit:
+
+	k_mutex_unlock(&s_spinel_hdlc_wr_lock);
+
+	if (error != OT_ERROR_NONE)
+	{
+		otLogCritPlat("error = 0x%x", error);
+	}
+
+	return error;
+}
+
+otError HdlcInterface::WaitForFrame(uint64_t aTimeoutUs)
+{
+	otError  error	  = OT_ERROR_RESPONSE_TIMEOUT;
```

Comment:
Alignment is off, please check this file and the header globally, generally there's no need for extra padding within the code.

Reply:
resolved